File: InheritanceTest/results/InheritanceTest.cs
using System;

public interface IPrinterBase
{
    int SomeBaseFunc();
}

public interface IPrinter: IPrinterBase
{
    void PrintIt();
}

public class BasePrinter: IPrinter
{
    public virtual string GetValue()
    {
        return "Base";
    }
    
    public void PrintIt()
    {
        Console.WriteLine($"BasePrinter: {this.GetValue()}");
    }
    
    public int SomeBaseFunc()
    {
        return 42;
    }
}

public class ChildPrinter: BasePrinter
{
    public override string GetValue()
    {
        return "Child";
    }
}

public class TestClass
{
    public IPrinter GetPrinter(string name)
    {
        var result = name == "child" ? new ChildPrinter() : new BasePrinter();
        return result;
    }
    
    public void TestMethod()
    {
        var basePrinter = this.GetPrinter("base");
        var childPrinter = this.GetPrinter("child");
        basePrinter.PrintIt();
        childPrinter.PrintIt();
        Console.WriteLine($"{basePrinter.SomeBaseFunc()} == {childPrinter.SomeBaseFunc()}");
    }
}

public class Program
{
    static public void Main()
    {
        try 
        {
            new TestClass().TestMethod();
        }
        catch (System.Exception e)
        {
            System.Console.WriteLine($"Exception: {e.Message}");
        }
    }
}

File: InheritanceTest/results/InheritanceTest.ts
interface IPrinterBase {
  someBaseFunc();
}

interface IPrinter extends IPrinterBase {
  printIt();
}

class BasePrinter implements IPrinter {
  getValue() {
    return "Base";
  }
  
  printIt() {
    console.log(`BasePrinter: ${this.getValue()}`);
  }
  
  someBaseFunc() {
    return 42;
  }
}

class ChildPrinter extends BasePrinter {
  getValue() {
    return "Child";
  }
}

class TestClass {
  getPrinter(name: string) {
    const result = name == "child" ? new ChildPrinter() : new BasePrinter();
    return result;
  }
  
  testMethod() {
    const basePrinter = this.getPrinter("base");
    const childPrinter = this.getPrinter("child");
    basePrinter.printIt();
    childPrinter.printIt();
    console.log(`${basePrinter.someBaseFunc()} == ${childPrinter.someBaseFunc()}`);
  }
}

try {
  new TestClass().testMethod();
} catch(e) {
  console.log(`Exception: ${e.message}`);
}